src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/DSA.cs

```diff
@@ -194,45 +541,371 @@ public virtual bool TrySignData(ReadOnlySpan<byte> data, Span<byte> destination,
+        ///   An error occurred in the verification operation.
+        /// </exception>
+        protected virtual bool VerifySignatureCore(
+            ReadOnlySpan<byte> hash,
+            ReadOnlySpan<byte> signature,
+            DSASignatureFormat signatureFormat)
+        {
+            // This method is expected to be overriden with better implementation
+
+            byte[]? sig = this.ConvertSignatureToIeeeP1363(signatureFormat, signature);
 
-            if (TryHashData(data, buf, hashAlgorithm, out int hashLength))
+            // If the signature failed normalization to IEEE P1363 then it
+            // obviously doesn't verify.
+            if (sig == null)
             {
-                hash = buf.Slice(0, hashLength);
+                return false;
             }
-            else
+
+            // The only available implementation here is abstract method, use it.
+            // Since it requires an exactly-sized array, skip pooled arrays.
+            return VerifySignature(hash, sig);
+        }
+
+        private ReadOnlySpan<byte> HashSpanToTmp(
+            ReadOnlySpan<byte> data,
+            HashAlgorithmName hashAlgorithm,
+            Span<byte> tmp)
+        {
+            Debug.Assert(tmp.Length == HashBufferStackSize);
+
+            if (TryHashData(data, tmp, hashAlgorithm, out int hashSize))
             {
-                // Use ArrayPool.Shared instead of CryptoPool because the array is passed out.
-                byte[] array = ArrayPool<byte>.Shared.Rent(data.Length);
-                try
-                {
-                    data.CopyTo(array);
-                    hash = HashData(array, 0, data.Length, hashAlgorithm);
-                }
-                finally
-                {
-                    Array.Clear(array, 0, data.Length);
-                    ArrayPool<byte>.Shared.Return(array);
-                }
+                return tmp.Slice(0, hashSize);
             }
 
-            return VerifySignature(hash, signature);
+            // This is not expected, but a poor virtual implementation of TryHashData,
+            // or an exotic new algorithm, will hit this fallback.
+            return HashSpanToArray(data, hashAlgorithm);
         }
 
-        public virtual bool VerifySignature(ReadOnlySpan<byte> hash, ReadOnlySpan<byte> signature) =>
-            VerifySignature(hash.ToArray(), signature.ToArray());
+        private byte[] HashSpanToArray(ReadOnlySpan<byte> data, HashAlgorithmName hashAlgorithm)
+        {
+            // Use ArrayPool.Shared instead of CryptoPool because the array is passed out.
+            byte[] array = ArrayPool<byte>.Shared.Rent(data.Length);
+            try
+            {
+                data.CopyTo(array);
+                return HashData(array, 0, data.Length, hashAlgorithm);
+            }
+            finally
+            {
+                Array.Clear(array, 0, data.Length);
+                ArrayPool<byte>.Shared.Return(array);
```

Comment:
It's fine to have the `Array.Clear` inside the _finally_ block, but general guidance is that we shouldn't return the array to the pool in exceptional cases since we should assume the buffer is in an indeterminate state. Let the GC reclaim it.